### ISSUE TITLE
Add a command to set the user role

### ISSUE THREAD

**Issue by 0xcodercrane** [ubiquity/ubiquibot#359]
## Problems to solve
From https://github.com/ubiquity/ubiquibot/issues/272
`The goal is to financially incentivize core contributors for their work beyond just the bounty hunter solving bounties.`

we should filter the users by its role. The original idea was to get the organization membership but getting the [Resource not accessible by integration](https://github.com/probot/probot/issues/1243) errors which has not been resolved yet.
 
## Proposal 
The alternative way is to get/set the user role using a new command like `/role ${username} ${role}`.
The role name needs to be one of the following roles compatible with [github organization membership](https://docs.github.com/en/rest/orgs/members?apiVersion=2022-11-28#get-organization-membership-for-a-user)

1. admin
2. member
3. collaborator

## Steps to do
1. create a new command like `/role ${username} ${role}`
2. Admins can only run this command
4. Store the role into `users` table of supabase database

**Comment by pavlovcik** [comment]
I delegate the spec check to @rndquu but the priority level seems ok to me.

**Comment by rndquu** [comment]
> we should filter the users by its role. The original idea was to get the organization membership

@0xcodercrane so in the "comment incentives" feature we need to check whether a user is a member of a particular organisation and that API request responds with `Resource not accessible by integration`, right?

It seems easier to try to fix the "Resource not accessible by integration" because:
1. We already use github organisation roles in the bot's code so adding another RBAC feature (which uses the same role names as github) will make things more complicated. It is easier to support either github roles either our own RBAC implementation.
2. Github already has the built-in RBAC feature, implementing it from scratch is a time-consuming task
3. Partner will have to (as far as I understand) use `/role USER_NAME USER_ROLE` for all org members which basically duplicates an existing github RBAC

So perhaps we can find a solution which uses "standard" github roles and API. Perhaps I can help with the "Resource not accessible by integration" error.

**Comment by 0xcodercrane** [comment]
> so in the "comment incentives" feature we need to check whether a user is a member of a particular organisation and that API request responds with Resource not accessible by integration, right?

Right 
1. I am wondering where we use it.
2. Would you post the reference link for the built-in RBAC feature? I am not sure I understood it correctly


TL;DR; Whatever is fine if you can find a solution for the error.  if you can do some research on it, appreciate it. we started building the bot with probot framework. so it would be better if you can find a solution that the probot supports

**Comment by rndquu** [comment]
> I am wondering where we use it

We use github roles in this [method](https://github.com/ubiquity/ubiquibot/blob/development/src/helpers/issue.ts#L153) to check whether a user can:
- edit labels
- set bounty hunter's multiplier
- use the `allow` command

> Would you post the reference link for the built-in RBAC feature?

I meant that we should use either [github ogranisation roles](https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/roles-in-an-organization) either implement our own solution for RBAC (described in this task)

> TL;DR; Whatever is fine if you can find a solution for the error. if you can do some research on it, appreciate it. we started building the bot with probot framework. so it would be better if you can find a solution that the probot supports

So what method exactly returns the `Resource not accessible by integration` error? This [one](https://github.com/ubiquity/ubiquibot/blob/development/src/helpers/user.ts#L35)?

**Comment by 0xcodercrane** [comment]
> We use github roles in this [method](https://github.com/ubiquity/ubiquibot/blob/development/src/helpers/issue.ts#L153) to check whether a user can:

I can apparently say its not a membership of the organization. what it returns is the permission of the repository like `admin`, `read`, `write`. When you get a chance, please try to run it. We couldn't never get the `admin`, `member`, `billing_manager` with that.

**Comment by 0xcodercrane** [comment]
> So what method exactly returns the Resource not accessible by integration error? This [one](https://github.com/ubiquity/ubiquibot/blob/development/src/helpers/user.ts#L35)?

Yes it is

**Comment by rndquu** [comment]
> > So what method exactly returns the Resource not accessible by integration error? This [one](https://github.com/ubiquity/ubiquibot/blob/development/src/helpers/user.ts#L35)?
> 
> Yes it is

I'll check it

**Comment by pavlovcik** [comment]
I am unsure where the original specification I wrote is or why it is not associated with this bounty, but I want to clarify that we should have delegated access control for the following actions:

- setting the `priority` level of the task - this is intended for project managers 
- setting the `time` estimate of the task - this is intended for technical architects to agree that the proposed solution is the most efficient to handle the business problem
- directly setting the `price` of the task - sometimes custom agreements are made with the bounty hunters, and the price is not a function of the priority or time of the task.
- `autopay` enable/disable.

What this means is that we need to be able to grant specific permissions to the contributors, and we will need to support 4^4 permutations of access control. I do not believe that this is possible with the built-in GitHub roles, as there are less than 16 options.

**Comment by rndquu** [comment]
> we should have delegated access control for the following actions: `autopay` enable/disable

This feature is about to be merged and it states that [only repo admin](https://github.com/ubiquity/ubiquibot/pull/363/files#diff-a977d1d34737ff4d3c0d836f5c3f40074bc74290d79d2087a3f52f9a4956ff23R30) can use the `autopay` command. Adding a separate access control of who can use the `autopay` command seems to be redundant at the current project stage, checking that the caller has the repo permission "admin" is enough for now.

**Comment by pavlovcik** [comment]
> checking that the caller has the repo permission "admin" is enough for now.

I agree with this. The required permutations to support is now 3^3=27.

**Comment by rndquu** [comment]
> > We use github roles in this [method](https://github.com/ubiquity/ubiquibot/blob/development/src/helpers/issue.ts#L153) to check whether a user can:
> 
> I can apparently say its not a membership of the organization. what it returns is the permission of the repository like `admin`, `read`, `write`. When you get a chance, please try to run it. We couldn't never get the `admin`, `member`, `billing_manager` with that.

Try:
1. Set the "members" permission to "read-only" under organisation permissions
2. Accept the updated permissions for the ubiquity org 

Everything worked fine for my local testing org with the "members" permission

**Comment by pavlovcik** [comment]
@0xcodercrane is this issue deprecated due to our current implementation for 
1. `/allow` https://github.com/ubiquibot/staging/issues/61#issuecomment-1637583271
2. setting `admin` or `billing_manager` roles?

**Comment by 0xcodercrane** [comment]
> @0xcodercrane is this issue deprecated due to our current implementation for
> 
> 1. `/allow` [QA for self config syntax ubiquibot/staging#61 (comment)](https://github.com/ubiquibot/staging/issues/61#issuecomment-1637583271)
> 2. setting `admin` or `billing_manager` roles?

The issue is different from the QA issue you're saying. It was kind of a permission issue of github app. I follow up @rndquu 's way here

**Comment by 0xcodercrane** [comment]
updated permissions of ubiquibot and ubiquibot-dev
